File: apps/control-plane/server/multiRailFailover.ts
export type Status = 'submitted'|'pending'|'settled'|'failed'|'held'|'unknown';
export type Intent = { id:string; idempotencyKey:string };
export type Submission = { status:Status; providerRef?:string; safeToRetry?:boolean };
export interface Rail { readonly name:string; submit(i:Intent):Promise<Submission>; query(i:Intent):Promise<Submission>; }
export class UnknownOutcome extends Error { constructor(message='provider outcome unknown; fallback prohibited'){super(message);this.name='UnknownOutcome';} }
export class MultiRailCoordinator { private readonly records=new Map<string,{rail:string;submission:Submission}>();
  async execute(i:Intent,primary:Rail,secondary:Rail):Promise<{rail:string;submission:Submission}>{
    if(!i.id||!i.idempotencyKey) throw new Error('intent and idempotency key required');
    const cached=this.records.get(i.idempotencyKey); if(cached)return cached;
    let first:Submission;
    try{first=await primary.submit(i);}catch{
      try{first=await primary.query(i);}catch{throw new UnknownOutcome();}
    }
    if(['submitted','pending','settled'].includes(first.status)) return this.record(i,primary.name,first);
    if(!first.safeToRetry || !['failed','held'].includes(first.status)) throw new UnknownOutcome();
    const second=await secondary.submit(i);
    if(!['submitted','pending','settled'].includes(second.status))throw new UnknownOutcome('secondary outcome not accepted');
    return this.record(i,secondary.name,second);
  }
  private record(i:Intent,rail:string,submission:Submission){const prior=this.records.get(i.idempotencyKey);if(prior)return prior;const r={rail,submission};this.records.set(i.idempotencyKey,r);return r;}
}

File: services/risk-compliance-core/multirail_failover.py
from dataclasses import dataclass
from enum import Enum
from threading import Lock
from typing import Protocol

class Status(str, Enum):
    SUBMITTED='submitted'; PENDING='pending'; SETTLED='settled'; FAILED='failed'; HELD='held'; UNKNOWN='unknown'
@dataclass(frozen=True)
class Intent: id:str; idempotency_key:str
@dataclass(frozen=True)
class Submission: status:Status; reference:str|None=None; safe_to_retry:bool=False
class Rail(Protocol):
    name:str
    def submit(self, intent:Intent)->Submission: ...
    def query(self, intent:Intent)->Submission: ...
class UnknownOutcome(Exception): pass
class Coordinator:
    def __init__(self): self._lock=Lock(); self._records:dict[str,tuple[str,Submission]]={}
    def execute(self,intent:Intent,primary:Rail,secondary:Rail)->tuple[str,Submission]:
        if not intent.id or not intent.idempotency_key: raise ValueError('intent and idempotency key required')
        with self._lock:
            if intent.idempotency_key in self._records:return self._records[intent.idempotency_key]
        try: first=primary.submit(intent)
        except TimeoutError:
            try: first=primary.query(intent)
            except Exception as exc: raise UnknownOutcome('primary outcome unknown') from exc
        if first.status in {Status.SUBMITTED,Status.PENDING,Status.SETTLED}: return self._record(intent,primary.name,first)
        if not (first.safe_to_retry and first.status in {Status.FAILED,Status.HELD}): raise UnknownOutcome('primary outcome unknown; fallback prohibited')
        second=secondary.submit(intent)
        if second.status not in {Status.SUBMITTED,Status.PENDING,Status.SETTLED}: raise UnknownOutcome('secondary outcome not accepted')
        return self._record(intent,secondary.name,second)
    def _record(self,i:Intent,name:str,s:Submission):
        with self._lock:
            return self._records.setdefault(i.idempotency_key,(name,s))
